sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/localstorage/LocalStorageTelemetryPipelineListener.java

```diff
@@ -71,21 +74,83 @@ private void processStatusCode206(TelemetryPipelineRequest request, TelemetryPip
         Set<ResponseError> errors = response.getErrors();
         errors.forEach(error -> logger.verbose("Error in telemetry: {}", error));
         if (!errors.isEmpty()) {
-            List<TelemetryItem> originalTelemetryItems = new ArrayList<>();
-            for (ByteBuffer byteBuffer : request.getByteBuffers()) {
-                originalTelemetryItems.addAll(deserialize(byteBuffer.array()));
-            }
-            List<TelemetryItem> toBePersisted = new ArrayList<>();
+            List<ByteBuffer> originalByteBuffers = request.getByteBuffers();
+            byte[] encodedBytes = convertByteBufferListToByteArray(originalByteBuffers);
+            byte[] decodedBytes = ungzip(encodedBytes); // decode is needed in order to split by newline correctly
+            List<byte[]> decodedByteArrayList = splitBytesByNewline(decodedBytes);
+            List<ByteBuffer> toBePersisted = new ArrayList<>();
             for (ResponseError error : errors) {
                 if (StatusCode.isRetryable(error.getStatusCode())) {
-                    toBePersisted.add(originalTelemetryItems.get(error.getIndex()));
+                    toBePersisted.add(ByteBuffer.wrap(decodedByteArrayList.get(error.getIndex())));
                 }
             }
+
             if (!toBePersisted.isEmpty()) {
                 localFileWriter.writeToDisk(
-                    request.getConnectionString(), serialize(toBePersisted), "Received partial response code 206");
+                    request.getConnectionString(), gzip(toBePersisted), "Received partial response code 206");
+            }
+        }
+    }
+
+    // convert list of byte buffers to byte array
+    private static byte[] convertByteBufferListToByteArray(List<ByteBuffer> byteBuffers) {
```

Comment:
(more descriptive name could help reading)
```suggestion
    private static byte[] concatenate(List<ByteBuffer> byteBuffers) {
```

Reply:
i prefer the old name. i've updated the comment.